scripts/ado-terraform-nagger.py

```diff
@@ -0,0 +1,399 @@
+    """
+    This function searches for a version number in the input `text`
+    using the specified regular expression `regex`.
+    If a match is found, the version number is returned as a string.
+    If no match is found, the function returns None.
+
+    Parameters:
+        text (str): The input text to be searched.
+        regex (str): The regular expression pattern used to search for the version.
+                     (regex must include "semver" named match group)
+
+    Returns:
+        str: The version number as a string if found, otherwise None.
+    """
+    matches = re.finditer(regex, text, re.MULTILINE | re.IGNORECASE | re.VERBOSE)
+    for match in matches:
+        return match.group("semver").strip()
+    else:
+        return None
+
+
+def terraform_version_checker(terraform_version):
+    # Load config file with pre-defined versions
+    try:
+        filename = "ado-terraform-nagger-versions.json"
+        global config
+        config = json.loads(load_file(filename))
+    except json.JSONDecodeError:
+        logger.error(f"{filename} contains invalid JSON")
+    except Exception as e:
+        logger.error(f"Error loading {filename}: {e}")
+
+    # Error if terraform version is lower than specified within config.
+    if version.parse(terraform_version) < version.parse(
+        config["terraform"]["error_below"]
+    ):
+        log_message(
+            slack_user_id,
+            slack_webhook_url,
+            "error",
+            f"Detected terraform version {terraform_version} "
+            f'is lower than {config["terraform"]["error_below"]}. '
+            "This is no longer allowed, please upgrade...",
+        )
+
+    # Warn if terraform version is lower than specified within config.
+    if version.parse(terraform_version) < version.parse(
+        config["terraform"]["warn_below"]
+    ):
+        log_message(
+            slack_user_id,
+            slack_webhook_url,
+            "warning",
+            f"Detected terraform version {terraform_version} "
+            f'is lower than {config["terraform"]["warn_below"]}. '
+            "Please upgrade...",
+        )
+
+
+def main():
+
+    # Retrieve HMCTS github to slack user mappings
+    hmcts_github_slack_user_mappings = get_hmcts_github_slack_user_mappings()
+    # Attempt to retrieve github username
+    github_user = os.getenv("BUILD_SOURCEVERSIONAUTHOR")
+    # Attempt to map github user to slack username
+    global slack_user_id
+    slack_user_id = get_github_slack_user_mapping(
+        hmcts_github_slack_user_mappings, github_user
+    )
+    if not slack_user_id:
+        log_message(None, None, "warning", "Missing slack user ID from github mapping")
+    # Atempt to retrieve slack webhook URL
+    global slack_webhook_url
+    slack_webhook_url = os.getenv("SLACK_WEBHOOK_URL")
+    if not slack_webhook_url:
+        log_message(None, None, "warning", "Missing slack webhook URL")
+
+    command = ["terraform", "version", "--json"]
```

Comment:
(optional for spike)

The ticket wasn't the clearest but one of the intentions for this was to enable the capability of using pipeline naggers across Jenkins and ADO without having to maintain it in two places.

I think this _should_ be refactored to:
a) have a pipeline nagger functionality
b) have a set of naggers (just one for now being the terraform one)

concretely I think this would mean a couple of things:
1. splitting the nagging parts of this out to a nagger file that's imported to do the notification part so that other naggers can use it.
2. not directly using ADO environment variables throughout and instead using an interface for it.

I think this may be worthwhile for either when a second nagger comes along or for when its integrated to Jenkins, but hopefully this adds more context.

Reply:
+1 to have a naggers file/repo that we can append to for other pieces